### ISSUE TITLE
[BUG] uid not generated for some dashboard URls

### ISSUE THREAD

**Issue by nicholasgibson2** [grafana/grafana#11186]
Read before posting: 

- Questions should be posted to https://community.grafana.com. Please search there and here on GitHub for similar issues before creating a new issue. 
- Checkout FAQ: https://community.grafana.com/c/howto/faq
- Checkout How to troubleshoot metric query issues: https://community.grafana.com/t/how-to-troubleshoot-metric-query-issues/50

Please include this information:

**Comment by marefr** [comment]
Can you please include more information. Please take one of these dashboards as example that has a unique title.

* What happens when you browse `http://<grafana url>/dashboard/db/<slug of dashboard title>`? Please include chrome dev tools network information if possible.
* What happens when you issue a curl or similar request to `http://<grafana url>/api/dashboards/db/<slug of dashboard title>`?

**Comment by nicholasgibson2** [comment]
@marefr 
- `https://<grafana url>/dashboard/db/<slug of dashboard title>` redirects to `https://<grafana url>/d/<slug of dashboard title>` then returns the standard 404 page. 
- `http://<grafana url>/api/dashboards/db/<slug of dashboard title>` returns a valid api response with all expected dashboard information. The uid field is blank in the response.

**Comment by marefr** [comment]
What database and version are you using for Grafana backend, sqlite (default), mysql or postgres? Would it be possible for you to do a `select * from migration_log` from backend database and include the result here? 

And for a sanity check, if you could select some rows from the dashboard table and have a look at the uid column, does it have null values or empty string?

Basically our migration script for database should have run the following (depends on database):
* Sqlite: `UPDATE dashboard SET uid=printf('%09d',id) WHERE uid IS NULL`
* Postgres: `UPDATE dashboard SET uid=lpad('' || id,9,'0') WHERE uid IS NULL`
* Mysql: `UPDATE dashboard SET uid=lpad(id,9,'0') WHERE uid IS NULL`

Thanks

**Comment by nicholasgibson2** [comment]
We are using mysql 5.6. The uid has null values for affected dashboards. Here are the log lines from our upgrade.

https://gist.github.com/nicholasgibson2/8375c46455565357327e967b65f50c20

**Comment by marefr** [comment]
What is the result if you execute `select lpad('' || id,9,'0') from dashboard`?

**Comment by nicholasgibson2** [comment]
```
mysql> select lpad('' || id,9,'0') from dashboard;
+----------------------+
| lpad('' || id,9,'0') |
+----------------------+
| 000000001            |
...
| 000000001            |
+----------------------+
180 rows in set (0.09 sec)
```
Also I noticed every column in dashboard has a count of 180 except uid which has a count of 174.

**Comment by marefr** [comment]
Do every row result in 000000001? It idea was that it should result in `00000000<dashboard id>`. 

> Also I noticed every column in dashboard has a count of 180 except uid which has a count of 174.
Can you elaborate what you mean by this thanks?

What happens when you execute `UPDATE dashboard SET uid=lpad(id,9,'0') WHERE uid IS NULL`?

**Comment by nicholasgibson2** [comment]
Yes, every row is 000000001. `select lpad(id,9,'0') from dashboard` shows your expected result. I found the cause of the issue though. We use a shared database for Grafana and one of our instances did not get updated. These dashboards were created or updated on the old version. Running `UPDATE dashboard SET uid=lpad(id,9,'0') WHERE uid IS NULL` creates a uid for these dashboard and fixes the issue. Closing this ticket as it is not a problem with Grafana.
